File: post_3/serving/fastapi-pickle/app/routes/predict.py
from fastapi import HTTPException, Request

from app.model import prepare_dataframe
from app.types import AdultIncomeInput, PredictionResponse


def predict(data: AdultIncomeInput, request: Request) -> PredictionResponse:
    model = request.app.state.model

    if not model.is_loaded():
        raise HTTPException(status_code=503, detail="Model not loaded")

    try:
        input_df = prepare_dataframe([data])

        predictions, probabilities = model.predict(input_df)

        predicted_class = int(predictions[0])
        probs = probabilities[0].tolist()

        income_label = ">50K" if predicted_class == 1 else "<=50K"

        return PredictionResponse(
            prediction=predicted_class,
            probabilities=probs,
            predicted_income=income_label,
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Prediction error: {str(e)}")


def predict_batch(data: list[AdultIncomeInput], request: Request):
    model = request.app.state.model

    if not model.is_loaded():
        raise HTTPException(status_code=503, detail="Model not loaded")

    if len(data) == 0:
        raise HTTPException(status_code=400, detail="Empty batch")

    try:
        input_df = prepare_dataframe(data)

        predictions, probabilities = model.predict(input_df)

        results = []
        for i in range(len(data)):
            income_label = ">50K" if predictions[i] == 1 else "<=50K"
            results.append(
                {
                    "prediction": int(predictions[i]),
                    "probabilities": probabilities[i].tolist(),
                    "predicted_income": income_label,
                }
            )

        return {"predictions": results, "batch_size": len(data)}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Batch prediction error: {str(e)}")

File: post_3/serving/axum-onnx/src/routes/predict.rs
use axum::{Json, extract::State, http::StatusCode};
use std::sync::Arc;
use crate::model::{Model, prepare_inputs};
use crate::types::{AdultIncomeInput, PredictionResponse, BatchPredictionResponse};

pub async fn predict(
    State(model): State<Arc<Model>>,
    Json(data): Json<AdultIncomeInput>,
) -> Result<Json<PredictionResponse>, (StatusCode, String)> {
    let inputs = prepare_inputs(&[data]).map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e))?;
    let (labels, probs) = model
        .predict(&inputs)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let prediction = labels[0];
    let probabilities = probs[0..2].to_vec();
    let predicted_income = if prediction == 1 { ">50K" } else { "<=50K" };

    Ok(Json(PredictionResponse {
        prediction,
        probabilities,
        predicted_income: predicted_income.to_string(),
    }))
}

pub async fn predict_batch(
    State(model): State<Arc<Model>>,
    Json(batch): Json<Vec<AdultIncomeInput>>,
) -> Result<Json<BatchPredictionResponse>, (StatusCode, String)> {
    if batch.is_empty() {
        return Err((StatusCode::BAD_REQUEST, "Empty batch".to_string()));
    }

    let inputs = prepare_inputs(&batch).map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e))?;
    let (labels, probs) = model
        .predict(&inputs)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let predictions = labels
        .iter()
        .enumerate()
        .map(|(i, &label)| {
            let probabilities = probs[i * 2..(i + 1) * 2].to_vec();
            let predicted_income = if label == 1 { ">50K" } else { "<=50K" };
            PredictionResponse {
                prediction: label,
                probabilities,
                predicted_income: predicted_income.to_string(),
            }
        })
        .collect();

    Ok(Json(BatchPredictionResponse {
        predictions,
        batch_size: batch.len(),
    }))
}
